templates/src/prettyprint.c.erb

```diff
@@ -1,111 +1,186 @@
+                pm_location_t *location = &cast-><%= field.name %>;
+                if (location->start == NULL) {
+                    pm_buffer_append_string(output_buffer, " ∅\n", 5);
+                } else {
+                    pm_buffer_append_byte(output_buffer, ' ');
+                    prettyprint_location(output_buffer, parser, location);
+                    pm_buffer_append_string(output_buffer, " = \"", 4);
+                    prettyprint_source(output_buffer, location->start, (size_t) (location->end - location->start));
+                    pm_buffer_append_string(output_buffer, "\"\n", 2);
+                }
             <%- when Prism::UInt32Field -%>
-            char <%= field.name %>_buffer[12];
-            snprintf(<%= field.name %>_buffer, sizeof(<%= field.name %>_buffer), "+%d", ((pm_<%= node.human %>_t *)node)-><%= field.name %>);
-            pm_buffer_append_str(buffer, <%= field.name %>_buffer, strlen(<%= field.name %>_buffer));
+                pm_buffer_append_format(output_buffer, " %d\n", cast-><%= field.name %>);
             <%- when Prism::FlagsField -%>
-            char <%= field.name %>_buffer[12];
-            snprintf(<%= field.name %>_buffer, sizeof(<%= field.name %>_buffer), "+%d", (uint32_t)(node->flags & ~PM_NODE_FLAG_COMMON_MASK));
-            pm_buffer_append_str(buffer, <%= field.name %>_buffer, strlen(<%= field.name %>_buffer));
+                bool found = false;
+                <%- found = flags.find { |flag| flag.name == field.kind }.tap { |found| raise "Expected to find #{field.kind}" unless found } -%>
+                <%- found.values.each do |value| -%>
+                if (cast->base.<%= field.name %> & PM_<%= found.human.upcase %>_<%= value.name %>) {
+                    if (found) pm_buffer_append_byte(output_buffer, ',');
+                    pm_buffer_append_string(output_buffer, " <%= value.name.downcase %>", <%= value.name.bytesize + 1 %>);
+                    found = true;
+                }
+                <%- end -%>
+                if (!found) pm_buffer_append_string(output_buffer, " ∅", 4);
+                pm_buffer_append_byte(output_buffer, '\n');
             <%- else -%>
             <%- raise -%>
             <%- end -%>
+            }
             <%- end -%>
-            pm_buffer_append_str(buffer, ")", 1);
+
             break;
         }
         <%- end -%>
     }
 }
 
-void
-pm_print_node(pm_parser_t *parser, pm_node_t *node) {
-    pm_buffer_t buffer;
-    if (!pm_buffer_init(&buffer)) return;
-
-    prettyprint_node(&buffer, parser, node);
-    printf("%.*s\n", (int) buffer.length, buffer.value);
-
-    pm_buffer_free(&buffer);
-}
-
 // Pretty-prints the AST represented by the given node to the given buffer.
 PRISM_EXPORTED_FUNCTION void
-pm_prettyprint(pm_parser_t *parser, pm_node_t *node, pm_buffer_t *buffer) {
-    prettyprint_node(buffer, parser, node);
+pm_prettyprint(pm_buffer_t *output_buffer, const pm_parser_t *parser, const pm_node_t *node) {
```

Comment:
Reminder that we need to merge this into CRuby before we merge it here, since CRuby uses the `pm_prettyprint` API already.

Reply:
Yeah for sure